libraries/BLE/src/BLEServiceImp.cpp

```diff
@@ -0,0 +1,387 @@
+ *
+ */
+
+#include "internal/ble_client.h"
+
+#include "BLEServiceImp.h"
+#include "BLECallbacks.h"
+#include "BLEUtils.h"
+#include "BLECharacteristicImp.h"
+
+bt_uuid_16_t BLEServiceImp::_gatt_primary_uuid = {BT_UUID_TYPE_16, BT_UUID_GATT_PRIMARY_VAL};
+
+bt_uuid_t *BLEServiceImp::getPrimayUuid(void)
+{
+    return (bt_uuid_t *)&_gatt_primary_uuid;
+}
+
+BLEServiceImp::BLEServiceImp(BLEService& service): 
+    BLEAttribute(service.uuid(), BLETypeService), 
+    _start_handle(0),
+    _end_handle(0xFFFF),
+    _cur_discover_chrc(NULL)
+{
+    memset(&_characteristics_header, 0, sizeof(_characteristics_header));
+    service.setServiceImp(this);
+}
+
+BLEServiceImp::BLEServiceImp(const bt_uuid_t* uuid): 
+    BLEAttribute(uuid, BLETypeService),
+    _start_handle(0),
+    _end_handle(0xFFFF),
+    _cur_discover_chrc(NULL)
+{
+    memset(&_characteristics_header, 0, sizeof(_characteristics_header));
+}
+
+BLEServiceImp::~BLEServiceImp()
+{
+    releaseCharacteristic();
+}
+
+
+int BLEServiceImp::addCharacteristic(BLEDevice& bledevice, BLECharacteristic& characteristic)
+{
+    BLECharacteristicImp* characteristicImp = this->characteristic(characteristic.uuid());
+    if (NULL != characteristicImp)
+    {
+        pr_info(LOG_MODULE_BLE, "%s-%d: Already exist",__FUNCTION__, __LINE__);
+        return BLE_STATUS_SUCCESS;
+    }
+    characteristicImp = new BLECharacteristicImp(characteristic, bledevice);
+    pr_debug(LOG_MODULE_BLE, "%s-%d",__FUNCTION__, __LINE__);
+    if (NULL == characteristicImp)
+    {
+        return BLE_STATUS_NO_MEMORY;
+    }
+    
+    BLECharacteristicNodePtr node = link_node_create(characteristicImp);
+    if (NULL == node)
+    {
+        delete[] characteristicImp;
+        return BLE_STATUS_NO_MEMORY;
+    }
+    link_node_insert_last(&_characteristics_header, node);
+    pr_debug(LOG_MODULE_BLE, "%s-%d",__FUNCTION__, __LINE__);
+    return BLE_STATUS_SUCCESS;
+}
+
+int BLEServiceImp::addCharacteristic(BLEDevice& bledevice, 
+                                     const bt_uuid_t* uuid, 
+                                     uint16_t handle, 
+                                     unsigned char properties)
+{
+    BLECharacteristicImp* characteristicImp = this->characteristic(uuid);
+    if (NULL != characteristicImp)
+    {
+        pr_info(LOG_MODULE_BLE, "%s-%d: Already exist",__FUNCTION__, __LINE__);
+        return BLE_STATUS_SUCCESS;
+    }
+    pr_debug(LOG_MODULE_BLE, "%s-%d:handle-%d",__FUNCTION__, __LINE__,handle);
+    characteristicImp = new BLECharacteristicImp(uuid, 
+                                                 properties, 
+                                                 handle, 
+                                                 bledevice);
+    if (NULL == characteristicImp)
+    {
+        return BLE_STATUS_NO_MEMORY;
+    }
+    
+    BLECharacteristicNodePtr node = link_node_create(characteristicImp);
+    if (NULL == node)
+    {
+        delete[] characteristicImp;
```

Comment:
This in undefined behaviour-- memory allocated using `new` should bee freed using `delete`, not `delete []`